Add support for games without `default game` (MTG)

maplemedley opened the pull request:
This adds support for games like Mineclonia, and any game xcompat adds support for. The only asterix is dye support, the pattern for dyes has been expanded for common dye mods, but if a new dye mod has different dye names those might not work without an update to the pattern.

maplemedley committed:
Add support for games without `default`

This adds support for games like Mineclonia, and any game xcompat adds
support for. The only asterix is dye support, the pattern for dyes has
been expanded for common dye mods, but if a new dye mod has different
dye names those might not work without an update to the pattern.

maplemedley commented:
Tested on Mineclonia and Minetest. Dyes work as well as before.

SwissalpS commented:
Thanks. Please cleanup the luacheck errors.

maplemedley commented:
Most of the errors are just for `xcompat` being undefined, but this is intended behaviour as this is a dependency. Other than that it was just a couple of whitespace errors, cleaned those up

maplemedley commented:
This patch _might_ clear out the warnings, will have to wait for the luacheck

Niklp09 commented:
Luacheck does not actually execute the code, so your patch is at the wrong place.

Just add this to the .luacheckrc file
```lua
read_globals = {
	"xcompat",
}
```

> will have to wait for the luacheck

You can also run luacheck on your device after installing it

maplemedley commented:
Thanks for the help! I hadn't heard about luacheck before, got it tested locally

wsor4035 commented on .luacheckrc:
```diff
@@ -10,4 +10,5 @@ read_globals = {
 	"player_monoids",
 	"pova",
 	"unifieddyes",
+        "xcompat",
```
please follow above nesting

wsor4035 reviewed (approved):
one minor nit pick

wsor4035 reviewed (approved):
lgtm

Niklp09 commented on crafts.lua:
```diff
@@ -39,13 +41,17 @@ local translated_colors = {
 	if not color then
-		color = string.match(name, "^dye:(.+)$")
+		color = string.match(name, dye_prefix_pattern_universal.."(.+)$")
+		string.gsub(name, dye_suffix_pattern_farlands, "")
```
Shouldn't this continue using `color`? Otherwise, line 54 wouldn't do anything.
```suggestion
		string.gsub(color, dye_suffix_pattern_farlands, "")
```

maplemedley replied:
Oh thanks for spotting that, without that line farlands wouldn't have color support

maplemedley committed:
Refactor & fix Farlands dye support

maplemedley commented:
Latest commit tested and working on Minetest, Mineclonia & Farland.

maplemedley committed:
Add some mcl colors, make names & their order consistent

maplemedley commented:
Latest commit tested and working on Minetest, Mineclonia and Farlands. 
I also noticed a pre-existing bug (tested and present on fresh ContendDB install) that causes dye crafting to only work in one quadrant of the crafting table, I'll see if I can fix it.

maplemedley committed:
Fix old crafting bug & refacfor crafting

maplemedley committed:
Minor refactor & reorder for clarity

maplemedley commented:
Fixed the pre-existing crafting bug, turned out that if the dye was placed before the hangglider in the crafting grid, the color from the hangglider was just re-applied. Refactored some things in the process of fixing this.

Color changing and repairing tested om Mineclonia, Minetest & Farlands

For Farlands only,  for some reason, every recipe repairs the item.

maplemedley committed:
Fix for Farlands `get_name` returning "" not nil

Farlands seems to return "" instead of nil/0 when no item is present.
By accounting for this, this commit fixes an issue where any empty slot
repaired the glider.

maplemedley commented:
Fixed the farlands repair issue & re-tested

SwissalpS commented on crafts.lua:
```diff
@@ -77,29 +120,51 @@ do
-			or "default:paper" == stack:get_name()
-		then
-			wear = 0
+		if name and name ~= "" then
```
this seems overkill. both nil and empty string don't equal to 'hangglider:hanglider'. Correct me if I'm wrong.

maplemedley replied:
It didn't work without `~= ""`

SwissalpS replied:
I see now, there could be trouble in the elseif and else block. I still think it would be nicer to have all the checks in the same indentation:
```lua
if not name or '' == name then
  --> exit/break/do nothing whatever applies
elseif 'hangglider:hangglider' == name then
  --> handle this case
elseif core.get_item_group(.......)  then
  --> handle that case
else
  --> handle rest cases
end
```
Maybe I'm still not seeing the full picture - it has been a while that I worked on this code.

maplemedley replied:
Not a bad idea, I used that method due to Lua's lack of `continue`, but an empty if block in the same level will do just as well. I also think the function's getting quite indented lol

SwissalpS replied:
luacheck may not like an empty block. Not sure what the directive is to silence that alert. Maybe a ```name = name``` or similar noop tricks luacheck.

SwissalpS replied:
OK, luacheck doesn't mind empty if-block, nvm.

maplemedley commented:
Applied @SwissalpS's idea to reduce indentation and re-tested on farlands

SwissalpS commented on crafts.lua:
```diff
@@ -64,9 +84,32 @@ local function get_color_name_from_color(color)
 
--- This recipe is just a placeholder
+
+-- Recipes ---------------------------------------------------------------------
```
nitpicking here: full width '---------' separator is a bit much especially since this is crafts.lua, we expect to find recipes here. Maybe the comment is more useful if it says: 'Repair Recipes'  or something to that effect. Basically the next comment says it.

maplemedley replied:
The reason is that the top half of the file before the header is dedicated to colors, the half beneath to recipes and recipe handlers

maplemedley replied:
It'd be good to have that separate anyways so trying that

maplemedley committed:
Separate color handling to new module

maplemedley commented:
Tested on farlands and minetest

SwissalpS commented:
That is a lot cleaner looking, thanks.

maplemedley committed:
Luacheck fix

wsor4035 commented:
luacheck issue?

SwissalpS commented on crafts.lua:
```diff
@@ -52,9 +52,9 @@ local function crafting_callback_handle_placeholder_recipe(crafted_item, _, old_
-					or minetest.get_item_group(name, string.match(repair_item, "^group:(.*)$")) ~= 0 
-				then 
+				if name == repair_item
+					or minetest.get_item_group(name, string.match(repair_item, "^group:(.*)$")) ~= 0
```
might as well use ``core`` here while the line is being changed anyway

SwissalpS commented on crafts.lua:
```diff
@@ -36,14 +36,14 @@ end
 		local name = stack:get_name()
 		if not name or name == "" then
-			-- This stack is empty, skip all checks
+			do end -- The stack is empty, do nothing and skip all checks for this stack.
```
try:
``do end -- luacheck: ignore``
and test that it only affects this line by e.g. doing ``local crafted_item = 'smth'`` on the next line.

SwissalpS replied:
otherwise push and pop methods may be necessary.

SwissalpS commented:
maybe there is a better ignore switch: https://luacheck.readthedocs.io/en/stable/inline.html

maplemedley committed:
Ignore empty `do end` luacheck warnings

maplemedley committed:
Replace `minetest.` with `core.`

maplemedley commented:
Tested on minetest & mineclonia
